_gen/lesson6.md

```diff
@@ -0,0 +1,896 @@
+# jsonnet -J lib -J vendor example4.jsonnet
+TRACE: vendor/testonnet/main.libsonnet:74 Testing suite example4.jsonnet
+{
+   "verify": "Passed 3 test cases"
+}
+
+// example1/example4.jsonnet.output
+~~~
+
+
+Testing `withImages()` is a bit more complex. In the library this function modifies the
+hidden `container::` field, which eventually gets added to the deployment in `new()`
+through late-initialization.
+
+On the left side again `new()` is called, this time `withImages()` is concatenated to get
+a deployment with an alternative image.
+
+On the right hand the container with name `httpd` in the deployment needs to be modified
+with the new image name, using the `mapContainerWithName` helper function to keep the
+test cases readable.
+
+Note that `mapContainerWithName` also preserves any other containers that may exist in
+the deployment, future-proofing the unit tests.
+
+### Test-driven development
+
+Let's write a test for a new function `webserver.withImagePullPolicy(policy)`, which can
+then be added as a feature to the library.
+
+~~~jsonnet
+local test = import 'testonnet/main.libsonnet';
+local webserver = import 'webserver/wrong1.libsonnet';
+
+local webserverName = 'webserver1';
+local base = import 'base.json';
+
+local mapContainerWithName(name, obj) =
+  {
+    local containers = super.spec.template.spec.containers,
+    spec+: { template+: { spec+: { containers: [
+      if c.name == name
+      then c + obj
+      else c
+      for c in containers
+    ] } } },
+  };
+
+
+test.new(std.thisFile)
++ test.case.new(
+  'Basic',
+  test.expect.eq(
+    webserver.new(webserverName),
+    base
+  )
+)
++ test.case.new(
+  'Change default replicas',
+  test.expect.eq(
+    webserver.new(webserverName, 2),
+    base { deployment+: { spec+: { replicas: 2 } } }
+  )
+)
++ test.case.new(
+  'Set alternative image',
+  test.expect.eq(
+    webserver.new(webserverName)
+    + webserver.withImage('httpd:2.5'),
+    base { deployment+: mapContainerWithName('httpd', { image: 'httpd:2.5' }) }
+  )
+)
++ test.case.new(
+  'Set imagePullPolicy',
+  test.expect.eq(
+    webserver.new(webserverName)
+    + webserver.withImagePullPolicy('Always'),
+    base { deployment+: mapContainerWithName('httpd', { imagePullPolicy: 'Always' }) }
+  )
+)
+
+// example1/example5.jsonnet
+~~~
+
+
+The new test 'Set imagePullPolicy' is very similar to 'Set alternative image'.
+
+To use the same `base`, `new()` is concatenated with
+`withImagePullPolicy('Always')` on the left.
+
+On the right it uses the `mapWithContainerName` helper to set `imagePullPolicy` on the
+`httpd` container.
+
+---
+
+~~~jsonnet
+local k = import 'k.libsonnet';
+local main = import 'main.libsonnet';
+
+main {
+  withImagePullPolicy(policy): {
+    container:
+      k.core.v1.container.withImagePullPolicy(policy),
+  },
+}
+
+// example1/lib/webserver/wrong1.libsonnet
+~~~
+
+
+Extending the library (referenced as `main`) with the `withImagePullPolicy()` function is
+quite straightforward.
```

Comment:
Being very nitpicky, you should first show what's the test output when you do not declare this function, otherwise, it isn't proper TDD :wink:

Reply:
Hypertechnically you are correct, respectfully going to ignore the nit.